### DISCUSSION TITLE
TIEMPO DE EJECUCCION Y RAM

### DISCUSSION THREAD

**Discussion by SantiagoOsorio3** [IIC2133-PUC/2024-2#24] [Tarea 0] [answered]
Hola, estaba leyendo las otras discussion y me acorde sobre lo del tiempo y la memoria del programa. En donde queria preguntar como especificamente puedo ver el tiempo exacto que demora mi programa y cuanta memoria ocupa. Esto porque lo hice de una manera pero me quiero asegurar...

Además, usando edd-docker que es lo que tengo que modificar en la carpeta de volume ? tengo que meter toda la tarea ? No me queda muy claro que hacer para poder correr valgrind con mi tarea. (con el ejemplo del setup pude, pero nose que debo cambiar para correr mi tarea).

Saludos.

**Comment by tomastrivino** [comment, marked as answer]
Puedes ver el tiempo que toma la ejecución de tu programa con el tomando `time` (sería algo como `time ./puffle input output`). Además, se encuentra en el enunciado una forma de ver cuánta memoria se ocupa: 
<img width="506" alt="image" src="https://github.com/user-attachments/assets/22533a92-c109-45a3-b037-cfbcf190f476">
Al ejecutar el último comando, se generará un archivo que puedes leer con el comando `ms_print massif.out`

Respecto a tu otra pregunta, me imagino que hiciste la tarea en otro entorno y que quieres probarla en el contenedor de Docker para usar `valgrind`. Lo que yo te recomendaría hacer es simplemente clonar el repositorio de tu tarea de Github y probarla en ese entorno. Si no queda claro o tienes más dudas, feliz de responderlas!

**Reply by SantiagoOsorio3** [reply]
Perfecto! probando el tiempo todo claro. Con respecto a lo otro, pego todo el repositorio, en donde hago make start desde edd-docker, luego cd a mi repositorio y dentro me deberia funcionar valgrind --tool=massif ?

**Reply by tomastrivino** [reply]
¿Hiciste la tarea en Mac? ¿Windows? sólo pa cachar bien qué es lo que buscas hacer. Yo me imagino que estás buscando probar en el contenedor tu tarea para usar Valgrind (y asumo que hiciste la tarea fuera del contenedor). Esto es lo que haría yo.
```
1.- Configurar el contenedor de Docker y ejecutarlo. Entrar a la terminal del contenedor
2.- Instalar git y poner mis credenciales
3.- Clonar mi repositorio de la tarea (ya debo haber hecho push antes con el progreso que tenía). Acceder a la carpeta
4.- make && valgrind ./puffle ...
```
La otra opción que tú dices es directamente copiar la carpeta de tu tarea dentro del contenedor (por ejemplo, en `volume/tarea`). En ese caso, te vas a dicha carpeta (con `cd`), ejecutas `make` y ya deberías poder usar los comandos como Valgrind.

**Reply by SantiagoOsorio3** [reply]
En MAC, perfecto ! ahi me quedo clarisimo, te cuento si tengo otra duda.

**Reply by SantiagoOsorio3** [reply]
Pude hacer lo de antes, sin embargo me sale esto...

root@9806622cfb78:~/volume/probar_setup/src# valgrind --leak-check=full --show-leak-kinds=all ./program/program input.txt output.txt
==217== Memcheck, a memory error detector
==217== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==217== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==217== Command: ./program/program input.txt output.txt
==217== 
==217== Invalid read of size 4
==217==    at 0x48B1D10: __vfscanf_internal (vfscanf-internal.c:345)
==217==    by 0x48B149B: __isoc99_fscanf (isoc99_fscanf.c:30)
==217==    by 0x108CCB: main (in /root/volume/probar_setup/src/program/program)
==217==  Address 0xc0 is not stack'd, malloc'd or (recently) free'd
==217== 
==217== 
==217== Process terminating with default action of signal 11 (SIGSEGV)
==217==  Access not within mapped region at address 0xC0
==217==    at 0x48B1D10: __vfscanf_internal (vfscanf-internal.c:345)
==217==    by 0x48B149B: __isoc99_fscanf (isoc99_fscanf.c:30)
==217==    by 0x108CCB: main (in /root/volume/probar_setup/src/program/program)
==217==  If you believe this happened as a result of a stack
==217==  overflow in your program's main thread (unlikely but
==217==  possible), you can try to increase the size of the
==217==  main thread stack using the --main-stacksize= flag.
==217==  The main thread stack size used in this run was 8388608.
==217== 
==217== HEAP SUMMARY:
==217==     in use at exit: 472 bytes in 1 blocks
==217==   total heap usage: 2 allocs, 1 frees, 944 bytes allocated
==217== 
==217== 472 bytes in 1 blocks are still reachable in loss record 1 of 1
==217==    at 0x4849D8C: malloc (in /usr/lib/aarch64-linux-gnu/valgrind/vgpreload_memcheck-arm64-linux.so)
==217==    by 0x48C9C97: __fopen_internal (iofopen.c:65)
==217==    by 0x108CAF: main (in /root/volume/probar_setup/src/program/program)
==217== 
==217== LEAK SUMMARY:
==217==    definitely lost: 0 bytes in 0 blocks
==217==    indirectly lost: 0 bytes in 0 blocks
==217==      possibly lost: 0 bytes in 0 blocks
==217==    still reachable: 472 bytes in 1 blocks
==217==         suppressed: 0 bytes in 0 blocks
==217== 
==217== For lists of detected and suppressed errors, rerun with: -s
==217== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
Segmentation fault



En donde investigando creo me di cuenta que es un problema esta relacionado a que se esta accediendo a una parte de memoria no permitida. Nose si es que me podrian explicar mas sobre esto..

**Reply by tomastrivino** [reply]
@SantiagoOsorio3  Te había respondido pensando que estabas ejecutando tu tarea. Sorry. Noté que estabas ejecutando un programa de prueba. 
Esos errores que surgen en Valgrind efectivamente indican que hay un problema con el manejo de la memoria (ya sea que la estás liberando y luego la intentas acceder, o que nunca la liberaste, o que accedes a una parte como `A[A.length]`, entre otros). ¿Pudiste solucionarlo?

**Reply by tomastrivino** [reply]
Un par de veces también he tenido problemas de ese tipo cuando me equivoco con el input (le doy un archivo que no existe). Eso también puede ocurrir
